File: hash/rust/rosetta-image-hash/src/internal/bitpack.rs
//! Boolean array ↔ hex string conversion. Row-major MSB-first, zero-padded.

use crate::ImageHashError;

pub fn pack(bits: &[Vec<bool>]) -> String {
    let h = bits.len();
    if h == 0 || bits[0].is_empty() {
        return String::new();
    }
    let w = bits[0].len();
    let total = h * w;
    let width = (total + 3) / 4;
    let mut bytes = vec![0u8; (total + 7) / 8];
    let mut bi = 0;
    for row in bits {
        for &b in row {
            if b {
                bytes[bi / 8] |= 1 << (7 - (bi % 8));
            }
            bi += 1;
        }
    }
    let mut out = String::with_capacity(width);
    for i in 0..width {
        let bit_pos = i * 4;
        let byte_idx = bit_pos / 8;
        let nibble = if bit_pos % 8 == 0 {
            bytes[byte_idx] >> 4
        } else {
            bytes[byte_idx] & 0x0F
        };
        out.push(char::from_digit(nibble as u32, 16).expect("nibble in 0..16"));
    }
    out
}

pub fn unpack_square(hex: &str) -> Result<Vec<Vec<bool>>, ImageHashError> {
    let bits = hex_to_bits(hex)?;
    let total = bits.len();
    let mut n: usize = 0;
    while n * n < total {
        n += 1;
    }
    if n * n != total {
        return Err(ImageHashError::InvalidHex(format!(
            "hex length {} ({total} bits) is not a square shape",
            hex.len()
        )));
    }
    let mut out: Vec<Vec<bool>> = Vec::with_capacity(n);
    let mut idx = 0;
    for _ in 0..n {
        let mut row = Vec::with_capacity(n);
        for _ in 0..n {
            row.push(bits[idx]);
            idx += 1;
        }
        out.push(row);
    }
    Ok(out)
}

pub fn unpack_flat(hex: &str, second_axis: usize) -> Result<Vec<Vec<bool>>, ImageHashError> {
    let bits = hex_to_bits(hex)?;
    let total = 14 * second_axis;
    if bits.len() < total {
        return Err(ImageHashError::InvalidHex(format!(
            "hex too short for 14x{} shape: {} bits",
            second_axis,
            bits.len()
        )));
    }
    let mut idx = bits.len() - total;
    let mut out: Vec<Vec<bool>> = Vec::with_capacity(14);
    for _ in 0..14 {
        let mut row = Vec::with_capacity(second_axis);
        for _ in 0..second_axis {
            row.push(bits[idx]);
            idx += 1;
        }
        out.push(row);
    }
    Ok(out)
}

fn hex_to_bits(hex: &str) -> Result<Vec<bool>, ImageHashError> {
    if hex.is_empty() {
        return Err(ImageHashError::InvalidHex("empty hex".to_string()));
    }
    for c in hex.chars() {
        if !c.is_ascii_digit() && !('a'..='f').contains(&c) {
            return Err(ImageHashError::InvalidHex(format!("invalid char {c:?} in {hex:?}")));
        }
    }
    let total = hex.len() * 4;
    let mut bits = vec![false; total];
    for (i, c) in hex.chars().enumerate() {
        let v = c.to_digit(16).expect("validated above");
        for b in 0..4 {
            bits[i * 4 + b] = (v >> (3 - b)) & 1 == 1;
        }
    }
    Ok(bits)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pack_4x4_pattern() {
        let bits: Vec<Vec<bool>> = vec![
            vec![true, false, true, false],
            vec![false, true, false, true],
            vec![true, true, true, true],
            vec![false, false, false, false],
        ];
        assert_eq!(pack(&bits), "a5f0");
    }

    #[test]
    fn pack_all_ones_8x8() {
        let bits: Vec<Vec<bool>> = (0..8).map(|_| vec![true; 8]).collect();
        assert_eq!(pack(&bits), "ffffffffffffffff");
    }

    #[test]
    fn pack_all_zeros_8x8() {
        let bits: Vec<Vec<bool>> = (0..8).map(|_| vec![false; 8]).collect();
        assert_eq!(pack(&bits), "0000000000000000");
    }

    #[test]
    fn unpack_square_is_inverse() {
        let expected: Vec<Vec<bool>> = vec![
            vec![true, false, true, false],
            vec![false, true, false, true],
            vec![true, true, true, true],
            vec![false, false, false, false],
        ];
        let got = unpack_square("a5f0").expect("unpack");
        assert_eq!(got.len(), 4);
        for y in 0..4 {
            assert_eq!(got[y], expected[y], "row {y}");
        }
    }

    #[test]
    fn unpack_flat_zeros() {
        let got = unpack_flat("00000000000", 3).expect("unpack");
        assert_eq!(got.len(), 14);
        assert_eq!(got[0].len(), 3);
        for y in 0..14 {
            for x in 0..3 {
                assert!(!got[y][x]);
            }
        }
    }

    #[test]
    fn unpack_square_non_square_errors() {
        assert!(unpack_square("12345").is_err());
    }

    #[test]
    fn unpack_invalid_chars_errors() {
        assert!(unpack_square("xyz!").is_err());
    }
}

File: hash/js/rosetta-image-hash/src/internal/bitpack.ts
/**
 * Boolean array ↔ hex string conversion. Row-major MSB-first, zero-padded
 * to ceil(M*N / 4) chars.
 *
 * Matches Python imagehash's _binary_array_to_hex which treats the bit string
 * as a big integer (RIGHT-aligned in the hex output). For bit counts that are
 * not multiples of 4, this prepends (4 - total%4)%4 leading zero bits so the
 * result agrees with Python's format (e.g. colorhash binbits=3 → 42 bits →
 * 11 hex chars with 2 leading zero bits prepended before the first nibble).
 */

import { ImageHashError } from "../hash.js";

export function pack(bits: boolean[][]): string {
  const h = bits.length;
  if (h === 0 || bits[0].length === 0) return "";
  const w = bits[0].length;
  const total = h * w;
  const width = (total + 3) >> 2; // ceil(total/4)
  // Number of leading zero-padding bits so the bit string is RIGHT-aligned
  // within the hex representation (matches Python's big-integer formatting).
  const pad = (4 - (total & 3)) & 3; // (4 - total%4) % 4
  const paddedTotal = total + pad;
  const byteCount = (paddedTotal + 7) >> 3; // ceil(paddedTotal/8)
  const bytes = new Uint8Array(byteCount);
  let bi = pad; // start writing after the leading zero bits
  for (const row of bits) {
    for (const b of row) {
      if (b) {
        bytes[bi >> 3] |= 1 << (7 - (bi & 7));
      }
      bi++;
    }
  }
  let out = "";
  for (let i = 0; i < width; i++) {
    const bitPos = i * 4;
    const byteIdx = bitPos >> 3;
    const nibble = (bitPos & 7) === 0 ? bytes[byteIdx] >> 4 : bytes[byteIdx] & 0x0f;
    out += nibble.toString(16);
  }
  return out;
}

export function unpackSquare(hex: string): boolean[][] {
  const bits = hexToBits(hex);
  const total = bits.length;
  let n = 0;
  while (n * n < total) n++;
  if (n * n !== total) {
    throw new ImageHashError(
      "InvalidHex",
      `hex length ${hex.length} (${total} bits) is not a square shape`,
    );
  }
  const out: boolean[][] = [];
  let idx = 0;
  for (let y = 0; y < n; y++) {
    const row: boolean[] = new Array(n);
    for (let x = 0; x < n; x++) {
      row[x] = bits[idx++];
    }
    out.push(row);
  }
  return out;
}

export function unpackFlat(hex: string, secondAxis: number): boolean[][] {
  const bits = hexToBits(hex);
  const total = 14 * secondAxis;
  if (bits.length < total) {
    throw new ImageHashError(
      "InvalidHex",
      `hex too short for 14x${secondAxis} shape: ${bits.length} bits`,
    );
  }
  let idx = bits.length - total; // align MSB-first if hex is wider than needed
  const out: boolean[][] = [];
  for (let y = 0; y < 14; y++) {
    const row: boolean[] = new Array(secondAxis);
    for (let x = 0; x < secondAxis; x++) {
      row[x] = bits[idx++];
    }
    out.push(row);
  }
  return out;
}

function hexToBits(hex: string): boolean[] {
  if (hex.length === 0) {
    throw new ImageHashError("InvalidHex", "empty hex");
  }
  for (const c of hex) {
    if (!/[0-9a-f]/.test(c)) {
      throw new ImageHashError("InvalidHex", `invalid char ${JSON.stringify(c)} in ${JSON.stringify(hex)}`);
    }
  }
  const total = hex.length * 4;
  const bits: boolean[] = new Array(total);
  for (let i = 0; i < hex.length; i++) {
    const v = parseInt(hex[i], 16);
    for (let b = 0; b < 4; b++) {
      bits[i * 4 + b] = ((v >> (3 - b)) & 1) === 1;
    }
  }
  return bits;
}
